### ISSUE TITLE
/saypreset not working with preset which has Sonos-Playlist

### ISSUE THREAD

**Issue by roelbroersma** [jishi/node-sonos-http-api#634]
Hello,

First of all, many thanks for the great work!

I noticed that when doing something like:

http://192.168.5.5:5005/office/saypreset/test/Hi%20there

And having a preset like:

**test.json**
{
  "players": [{ "roomName": "Office", "volume": 20}],
  "playlist": "Toilet"
}

The say is only being said but there is no playlist loaded and played.
Can someone please reproduce?

**Comment by jishi** [comment]
The saypreset is deliberately only using the grouping info and ignoring
playlist specifications, because it is supposed to resume whatever was
playing on the players after the announcement.

**Comment by roelbroersma** [comment]
OK,  So,  what is best for doing what I want:

- Say: "Hi There" (using TTS)
- Play Sonos playlist: "Toilet" (shuffled)

In 1 call ?   I can't do this by a preset, I can't do it by saypreset,   I can't even do it by clippreset (if i record the text into a clip).   So, what would be the solution then?

Thanks for your support!

**Comment by roelbroersma** [comment]
@jishi Or do I need to make a custom call/hook for this?

**Comment by jishi** [comment]
To do it in one call, you would need to do a custom hook for it. The other option (atm) would be to have a preset which has `state: stopped`, with your playlist, where you first do:

`preset/your_preset`
`saypreset/your_preset/hi there`
`any_player_in_preset/play` 

That will pre-load the group you want to say something in, and preload the source, then make the announcement, and then start the playback once restored. 

What you are asking are similar to https://github.com/jishi/node-sonos-http-api/issues/160, where you want a pre-announcement for a preset. Personally I don't really see a big use case for it, more than a gimmick, but maybe you can sell it for me :)

**Comment by roelbroersma** [comment]
@jishi   yes,  It is eaxactly the same as in #160 by @psmith3     I saw that issue before posting mine also.

Imagine:

1)
A user walks into a Hotel room,  as soon as he opens the door  (or when he first enters his keycard) he hears a voice: "Welcome at the room.... some explanation..." and then a playlist is played for background music.

2)
Same can happen for when a user enters another room, bathroom or toilet.
In my case I have something like: "Hello xxxxx ,.. you can ask me anything, just like: Alexa, skip track or  Alexa, what is the weather,  Meanwhile I will start a playlist"
(this works with bluetooth beacons and wifi to see if there are guests in the house)
I also have a push button on the wall for the cleaning personel.  When that button is pressed, the toilets and bathroom go into maintenance mode (1x automatic flushing and then stop) and a voice says some explanation and the radio is played on the background).


3)
Another could be when a user is walking in the night in a parking garage or outside.  Just to calm them down, you can have a voice saying "Welcome  to the garage mister ...... (based on license plate scanning), meanwhile we will play a playlist".
(and of course this playlist is always shuffled, otherwise you will hear the same song every time you enter the parking garage)


I think that a lot of people who use node-sonos-http-api use Domoticz, a Gira homeserver or something similiar to fit it in their homes.  It is just being used as an easy 'bridge'/connector/gateway between Sonos and other systems.

In #160  @psmith3 writes a preset example,  but this is not possible, is it?  (see below)
_{
	"players": [{
			"roomName": "Kitchen",
			"volume": 25
		},
		{
			"roomName": "Den",
			"volume": 20
		}
	],
	"say": "Taylor Swift Radio",
	"playMode": {
		"shuffle": true,
		"crossfade": true
	},
	"pauseOthers": true,
	"favorite": "Taylor Swift Radio"
}_

**Comment by jishi** [comment]
No, that example is not possible as of now, but it would make more sense to solve it that way rather than the opposite (saypreset starting playback).

The say logic is a very hard task, so I can see some benefit for it to be part of a preset like that because it's gonna be pretty fragile if you start combining multiple requests to achieve what you want. I'll have to think about it a bit. If you want to give it a try, you could try to make a custom action for it. I think most pieces needed for it is somewhat already built, but might not have been separated well enough.

**Comment by roelbroersma** [comment]
@jishi,  Did anything come to your mind about this?

**Comment by jishi** [comment]
@roelbroersma As you can see, #160 has been around for quite a while. It's not so much a matter of how to fix it, but rather having time to fix it. Everything that is achievable with multiple actions, are by default a very low priority, especially since the calls do wait for the action to finish, making them easily chainable in a script.

Hopefully I will have more time to work on this codebase this fall, but I can not make any promises to what and when stuff will be done.

**Comment by Khengsoon** [comment]
> To do it in one call, you would need to do a custom hook for it. The other option (atm) would be to have a preset which has `state: stopped`, with your playlist, where you first do:
> 
> `preset/your_preset`
> `saypreset/your_preset/hi there`
> `any_player_in_preset/play`
> 
> That will pre-load the group you want to say something in, and preload the source, then make the announcement, and then start the playback once restored.
> 
> What you are asking are similar to #160, where you want a pre-announcement for a preset. Personally I don't really see a big use case for it, more than a gimmick, but maybe you can sell it for me :)

Hi Jishi, I tried the new version, could you give me a guide how can I broadcast message to all or selected sonos speaker? I try to find the documentation but I could not find it. 

Thanks in advance
